### PR TITLE
state: correct FIRST_WRITE

### PR DESCRIPTION
**Pull request by daixiang0**
Signed-off-by: Long Dai <long0dai@foxmail.com>

# Description

As per discussion, if the etag value is null or is omitted, dapr handles the request in last write wins pattern. But since this may result in some complexity for saving state for a non-existing key, FIRST_WRITE option should be given priority over etag's null value, for handling the concurrency for the request.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #994

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by daixiang0** [comment]
ping @artursouza @yaron2

**Comment by daixiang0** [comment]
@pkedy @artursouza code finish, please help check CI fail, not sure it is env issue since no changes about it, thanks.

File: state/mongodb/mongodb.go [hunk 1]
```diff
@@ -179,6 +179,8 @@ func (m *MongoDB) setInternal(ctx context.Context, req *state.SetRequest) error
 	filter := bson.M{id: req.Key}
 	if req.ETag != nil {
 		filter[etag] = *req.ETag
+	} else if req.Options.Concurrency == state.FirstWrite {
+		filter[etag] = uuid.NewString()
```
Comment (artursouza) [inline state/mongodb/mongodb.go:183]: Was this tested? Why matching eTag on a random UUID fixes the firstWrite wins?

File: tests/conformance/state/state.go [hunk 2]
```diff
@@ -474,4 +474,72 @@ func ConformanceTests(t *testing.T, props map[string]string, statestore state.St
 		features := statestore.Features()
 		assert.False(t, state.FeatureETag.IsPresent(features))
 	}
+
+	if config.HasOperation("first-write") {
```
Comment (artursouza) [inline tests/conformance/state/state.go:478]: This is great! Thanks. BTW, can you add a scenario with first-write and an etag to update an item? Something like:

+ Write an item
+ Read item and get eTag (eTag1)
+ Set with FIRST_WRITE and eTag as eTag1
+ Read item again and check that eTag (eTag2 != eTag1) and value changes
+ Set again with FIRST_WRITE and eTag as eTag1
+ Check the set above failed.

**Review by artursouza — CHANGES_REQUESTED** [review]

File: state/mongodb/mongodb.go (hunk 1 above)
Comment (daixiang0) [reply state/mongodb/mongodb.go:183]: Tested, if set etag when firstWrite, then the second write would raise a error.

File: tests/conformance/state/state.go (hunk 2 above)
Comment (daixiang0) [reply tests/conformance/state/state.go:478]: Done.

**Comment by daixiang0** [comment]
ping @artursouza

**Review by artursouza — APPROVED** [review]

**Pull request merged**